alf/algorithms/data_transformer.py

```diff
@@ -0,0 +1,579 @@
+
+    def transform_timestep(self, time_step, state):
+        if self._stack_size == 1:
+            return time_step
+
+        is_first = time_step.step_type == StepType.FIRST
+        steps = state.steps + 1
+        steps[is_first] = 0
+        stack_axis = self._stack_axis
+        if stack_axis >= 0:
+            stack_axis += 1
+        first_samples = is_first.nonzero()
+
+        prev_frames = copy.copy(state.prev_frames)
+
+        def _stack_frame(obs, i):
+            prev_frames[i] = copy.copy(prev_frames[i])
+            # repeat the first frame
+            if first_samples.numel() > 0:
+                for t in range(self._stack_size - 1):
+                    # prev_frames[i][t] might be used somewhere else, we should
+                    # not directly modify it.
+                    prev_frames[i][t] = prev_frames[i][t].clone()
+                    prev_frames[i][t][first_samples] = obs[first_samples]
+            if obs.ndim > 1:
+                stacked = torch.cat(prev_frames[i] + [obs], dim=stack_axis)
+            else:
+                stacked = torch.stack(prev_frames[i] + [obs], dim=1)
+            prev_frames[i].pop(0)
+            prev_frames[i].append(obs)
+            return stacked
+
+        observation = time_step.observation
+        for i, field in enumerate(self._fields):
+            observation = alf.nest.utils.transform_nest(
+                observation, field, partial(_stack_frame, i=i))
+
+        return (time_step._replace(observation=observation),
+                FrameStackState(steps=steps, prev_frames=prev_frames))
+
+    def transform_experience(self, experience: Experience):
+        if self._stack_size == 1:
+            return experience
+
+        assert experience.batch_info != ()
+        batch_info: BatchInfo = experience.batch_info
+        replay_buffer: ReplayBuffer = experience.replay_buffer
+
+        with alf.device(replay_buffer.device):
+            # [B]
+            env_ids = convert_device(batch_info.env_ids)
+            # [B]
+            positions = convert_device(batch_info.positions)
+
+            prev_positions = torch.arange(self._stack_size -
+                                          1) - self._stack_size + 1
+
+            # [B, stack_size - 1]
+            prev_positions = positions.unsqueeze(
+                -1) + prev_positions.unsqueeze(0)
+            episode_begin_positions = replay_buffer.get_episode_begin_position(
+                positions, env_ids)
+            # [B, 1]
+            episode_begin_positions = episode_begin_positions.unsqueeze(-1)
+            # [B, stack_size - 1]
+            prev_positions = torch.max(prev_positions, episode_begin_positions)
+            # [B, 1]
+            env_ids = env_ids.unsqueeze(-1)
+            assert torch.all(
+                prev_positions[:, 0] >= replay_buffer.get_earliest_position(
+                    env_ids)
+            ), ("Some previous posisions are no longer in the replay buffer")
```

Comment:
What's the possible case of this assertion error?

Reply:
This can happen if num_ignore_earliest_frames is not correctly set for ReplayBuffer